share/p2p/middleware.go

```diff
@@ -0,0 +1,30 @@
+package p2p
+
+import (
+	"sync/atomic"
+
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/libp2p/go-libp2p/core/network"
+)
+
+var log = logging.Logger("shrex/middleware")
+
+func RateLimitMiddleware(inner network.StreamHandler, concurrencyLimit int) network.StreamHandler {
+	var parallelRequests int64
+	limit := int64(concurrencyLimit)
+	return func(stream network.Stream) {
+		current := atomic.AddInt64(&parallelRequests, 1)
+		defer atomic.AddInt64(&parallelRequests, -1)
+
+		if current > limit {
+			log.Debug("concurrency limit reached")
+			err := stream.Close()
+			if err != nil {
+				log.Errorw("server: closing stream", "err", err)
+			}
+			return
+		}
```

Comment:
I think we had a special error that was supposed to be sent in this case and not just stream closage

Reply:
We did, but we want to rethink this design - this would still require the overloaded nodes to read the request from the client, and then send a response - it is much better if we can just close the stream gracefully, and have the client interpret it as a rate limit response

Reply:
@Wondertan Can we merge this for now? We won't merge it until we have your okay, but we do want it in the feature branch. It significantly improves metrics

Reply:
As far as I remember initially @Wondertan purposed to Close the stream without response, but we revised it later with idea of sending response with status instead. Motivation was that it would allow us to know if stream was unintentionally closed or it was expected rate limiting. 

With closer look it turned out we can identify graceful stream close with error handling check, so there is no need to send extra message.

Pros of getting rid of extra message:
With explicit message approach we would have an extra round trip. Server that is already exhausted, would need to read request first, which will be thrown away. And then it would need to send response containing "Status_REFUSED". We could achieve the same, without extra round trip.

Also we can have rate limiting protocol agnostic, as it would not need to read out protocol specific request.

Reply:
The reasoning makes sense to me. Thanks for explaining. As long as there is explicit handling for the client side of this case I am ok with this approach. We will also need to update the specs in the Adr right after.